SukiUI/Utilities/Effects/EffectDrawBase.cs

```diff
@@ -51,30 +53,54 @@ public bool AnimationEnabled
         
         protected float AnimationSeconds => (float)_animationTick.Elapsed.TotalSeconds;
         
-        private readonly Stopwatch _animationTick = Stopwatch.StartNew();
+        private readonly Stopwatch _animationTick = new();
 
-        protected EffectDrawBase(Rect bounds)
+        protected EffectDrawBase()
         {
-            Bounds = bounds;
             var sTheme = SukiTheme.GetInstance();
             sTheme.OnBaseThemeChanged += v => ActiveVariant = v;
             ActiveVariant = sTheme.ActiveBaseTheme;
             sTheme.OnColorThemeChanged += t => ActiveTheme = t;
             ActiveTheme = sTheme.ActiveColorTheme!;
-            
         }
 
-        public void Render(ImmediateDrawingContext context)
+        public override void OnRender(ImmediateDrawingContext context)
         {
             var leaseFeature = context.TryGetFeature<ISkiaSharpApiLeaseFeature>();
             if (leaseFeature is null) throw new InvalidOperationException("Unable to lease Skia API");
             using var lease = leaseFeature.Lease();
-            var rect = SKRect.Create((float)Bounds.Width, (float)Bounds.Height);
-            if(lease.GrContext is null || ForceSoftwareRendering) // GrContext is null whenever
+            var rect = SKRect.Create((float)EffectiveSize.X, (float)EffectiveSize.Y);
+            if(lease.GrContext is null || ForceSoftwareRendering) // GrContext is null whenever there is no hardware acceleration available
                 RenderSoftware(lease.SkCanvas, rect);
             else
                 Render(lease.SkCanvas, rect);
         }
+        
+        public override void OnMessage(object message)
+        {
+            if (message == StartAnimations)
+            {
+                AnimationEnabled = true;
+                RegisterForNextAnimationFrameUpdate();
+            }
+            else if (message == StopAnimations)
+            {
+                AnimationEnabled = false;
+            }
+            else if (message is SukiEffect effect)
+            {
+                Effect = effect;
+            }
+        }
+
+        public override void OnAnimationFrameUpdate()
+        {
+            if (!AnimationEnabled) return;
+            Invalidate(GetRenderBounds());
```

Comment:
Unless you need to invalidate only small portion of the visual, you can just call `Invalidate()` method. It will reduce some overhead on calculating dirty rects per visual.